Always create usercases for web users

esoergel opened the pull request:
## Product Description


## Technical Summary
https://dimagi.atlassian.net/browse/USH-6180
Always create usercases for web users.  Also does some cleanup and unification that aligns with that.

## Feature Flag


## Safety Assurance


### Safety story


### Automated test coverage


### QA Plan
https://dimagi.atlassian.net/browse/QA-1217

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Release usercases for web users flag

Don't remove the flag yet - still useful for the migration script to
know which domains should already have these populated

esoergel committed:
Unify sync usercases implementations

esoergel committed:
Create user the normal way

The raw JSON was invalid - the users were missing domain memberships

esoergel committed:
Make sync code more fault-tolerant

I don't love being this defensive, but we have a million tests that use
partially provisioned user objects, and I think things that try to use a
CouchUser that's not wrapped properly

esoergel committed:
Update signal tests for new unified logic

Jtang-1 commented on corehq/apps/app_manager/tasks.py:
```diff
@@ -28,11 +28,7 @@
-    else:
-        users = CommCareUser.by_domain(domain_name)
-    for user in users:
+    for user in CouchUser.by_domain(domain_name):
```
both are better options than what I proposed, and ya using a dbaccessor does not sense. I think you can use `get_all_user_rows`

Jtang-1 replied:
went ahead and made the change [d6f9c1f](https://github.com/dimagi/commcare-hq/pull/36901/commits/d6f9c1f7cd058a1d76c6801613c33b9af32ee1d0)

esoergel replied:
Thanks!  I took a look at `get_all_user_rows` and it looks like it still does just 1-4 calls to `db.view`, so it still results in large DB hits.  There are some other utils that use `iter_docs` to fetch in chunks, so I copied those examples.

Jtang-1 replied:
I see, thanks i didn't consider that. 

And just for my understanding, I see your new commit still ends up using `get_all_user_rows`, so it still does 1-4 calls to `db.view`? But the memory gain here is that we're not fetching the entire docs, just the ids? So that we can then fetched/load the docs in memory in chunks.

Jtang-1 committed:
bug: sync_usercases expect the user to be wrapped correctly as either a WebUser or CommcareUser.

This change also streams the users instead of pulling it all into memory

esoergel committed:
Use iter_docs to iterate over all users on domain

When used with include_docs, get_all_user_rows still does a single
request to couch per doc type (`db.view` isn't a generator)

Jtang-1 commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -45,29 +44,22 @@ def calculate_indicators():
+    if settings.UNIT_TESTING and not domain_obj:
         return
-    if (user.project.call_center_config.enabled or user.project.usercase_enabled):
+    if (domain_obj.call_center_config.enabled or domain_obj.usercase_enabled):
```
if `domain_obj` does end up being None, wouldn't this error? I think we would just return early and not do anything in that case?

esoergel replied:
That's right, this would error.  Same goes for the old code though, so I _think_ we should be okay (`user.project` just calls `Domain.get_by_name`, as did the separate web user version).  If it fails, that means there's a real data issue - a user is a member of a domain that doesn't exist, so I'm kinda tempted to keep it in as a hard failure.

Though I do wonder if there's some way this could cause problems when we roll it out more broadly, like a leftover domain membership from a deleted domain or something.  What do you think?

Jtang-1 replied:
the migration actually doesn't go through this code path. It iterates through the domains and gets the users from each domains. As opposed to iterating through all users and getting the domain each user is a member of.

So I don't think it would run into any issues if the user doc has bad data

Jtang-1 committed:
Merge branch 'master' into es/webusercases
